src/hotspot/share/opto/callnode.cpp

```diff
@@ -1055,6 +1045,32 @@ bool CallStaticJavaNode::cmp( const Node &n ) const {
   return CallJavaNode::cmp(call);
 }
 
+Node* CallStaticJavaNode::Ideal(PhaseGVN* phase, bool can_reshape) {
+  CallGenerator* cg = generator();
+  if (can_reshape && cg != NULL) {
+    assert(IncrementalInlineMH, "required");
+    assert(cg->call_node() == this, "mismatch");
+    assert(cg->is_mh_late_inline(), "not virtual");
```

Comment:
Why you are expecting only MH calls uses CallStaticJavaNode?

Reply:
Late inlining already hits node limits regularly and `live_nodes() > LiveNodeCountInliningCutoff` is used to decide when to stop (see [Compile::inline_incrementally()](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/compile.cpp#L1907) for more details).

Reply:
Only `LateInlineMH` and `LateInlineVirtual` generators link themselves to call nodes. The former uses `CallStaticJavaNode` and the latter uses `CallDynamicJavaNode`.